### ISSUE TITLE
How do we support consumer groups in kinesis-spring integration

### ISSUE THREAD

**Issue by sidath80** [spring-cloud/spring-cloud-stream-binder-aws-kinesis#51]
How do we ensure that deliver message only single consumer in HA(High availability environment).

**Comment by artembilan** [comment]
A `single consumer` for the message is ensured by the `consumer group` abstraction in the Spring Cloud Stream:

https://docs.spring.io/spring-cloud-stream/docs/Elmhurst.RELEASE/reference/htmlsingle/#consumer-groups

https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/blob/master/spring-cloud-stream-binder-kinesis-docs/src/main/asciidoc/overview.adoc#kinesis-binder-overview

So, if you have several consumer instances, you should ensure to use the same `spring.cloud.stream.bindings.<channelName>.group` for all of them to let only one instance to process a message.

At the same time you should ensure a shared `MetadataStore` for all them to let their `Checkpointers` to do the proper parking and filtering logic.

By default Kinesis Binder uses `DynamoDbMetaDataStore`, which should rely on the same shared table between instances of your consumer.

Do we miss anything here ?

Thanks

**Comment by artembilan** [comment]
When we produce message, the group is ignored, we just use a `streamName` to put to Kinesis.
The `consumerGroup` makes sense only on the consumer side. So, several different groups can read the same message. 

Otherwise your concern isn't clear.

Please, elaborate.

**Comment by artembilan** [comment]
The `[CONSUMER_GROUP]:[STREAM]:[SHARD_ID]` is a pattern for `MetadataStore` keys: 
https://docs.spring.io/spring-integration/docs/5.0.5.RELEASE/reference/html/system-management-chapter.html#metadata-store

https://github.com/spring-projects/spring-integration-aws/blob/master/src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java#L615

You shound't worry about that. Only what you need to configure is a Spring Cloud Stream `group` option for the consumer definition: https://docs.spring.io/spring-cloud-stream/docs/Elmhurst.RELEASE/reference/htmlsingle/#binding-properties

Everything rest is done by the Framework automatically.

**Comment by sidath80** [comment]
Sorry I have couple of more questions,

As I Know producer just send data to the stream. not for the SHARD_ID. Is it possible to send directly to the SHARD_ID??

How consumer get saved data from the kinesis.Currently I am getting the latest data when producer produce and consumer will receive it.Is it possible to get the saved data (old data) ??

Thanks again.

**Comment by artembilan** [comment]
> Is it possible to send directly to the SHARD_ID??

Yes, you need to configure a `partitionKeyExpression` for the producer binding: https://docs.spring.io/spring-cloud-stream/docs/Elmhurst.RELEASE/reference/htmlsingle/#_producer_properties

Not sure though how to be able to know the `SHARD_ID`. Please, refer to AWS Kinesis documentation.

> Is it possible to get the saved data (old data) ??

For this purpose you need to configure a `shardIteratorType` on the consumer binding: https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/blob/master/spring-cloud-stream-binder-kinesis-docs/src/main/asciidoc/overview.adoc.

Sorry, do you read docs before coming to us with questions?..

**Comment by artembilan** [comment]
May we close it now?

Also consider to ask question in the future on the StackOverflow, GH is for bug reports and feature requests.

Thanks for understanding.

**Comment by Venkat2811** [comment]
Good explanation on Consumer Groups !

Maybe provide this in this documentation itself ? https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/blob/master/spring-cloud-stream-binder-kinesis-docs/src/main/asciidoc/overview.adoc

Since the project is relatively new and kinesis behavior is different from Kafka, providing this here (as part of binder doc) helps :)

Thanks

**Comment by artembilan** [comment]
@Venkat2811 ,

Thank you for feedback!

Would you mind to raise a Pull Request for Docs you are mentioning here?

**Comment by Venkat2811** [comment]
Sure @artembilan , will do !
